File: 1930-unique-length-3-palindromic-subsequences/1930-unique-length-3-palindromic-subsequences.java
class Solution {
    public int countPalindromicSubsequence(String s) {
        Integer[] first = new Integer[26];
        Integer[] last = new Integer[26];

        int result = 0;

        int index = 0;
        for (char c : s.toCharArray()) {
            if (first[c - 'a'] == null) {
                first[c - 'a'] = index;
            } else {
                last[c - 'a'] = index;
            }
            index++;
        }

        for (int i = 0; i < 26; i++) {
            if (first[i] == null || last[i] == null)
                continue;

            HashSet<Character> set = new HashSet<>();

            for (int j = first[i] + 1; j < last[i]; j++) {
                set.add(s.charAt(j));
            }

            result += set.size();
        }

        return result;
    }
}

File: 1930-unique-length-3-palindromic-subsequences/1930-unique-length-3-palindromic-subsequences.php
class Solution {

    /**
     * @param String $s
     * @return Integer
     */
    function countPalindromicSubsequence($s) {
        $ans = 0;
        $n = strlen($s);

        // Array to store first and last occurrences of each character
        $positions = [];

        for ($i = 0; $i < $n; $i++) {
            if (!isset($positions[$s[$i]])) {
                $positions[$s[$i]] = [$i, $i];
            } else {
                $positions[$s[$i]][1] = $i;
            }
        }

        foreach ($positions as $ch => $range) {
            $left = $range[0];
            $right = $range[1];

            if ($right - $left <= 1) {
                continue; // No valid subsequences in this range
            }

            $uniqueChars = [];
            for ($i = $left + 1; $i < $right; $i++) {
                $uniqueChars[$s[$i]] = true; // Store unique characters
            }

            $ans += count($uniqueChars); // Count the unique characters
        }

        return $ans;
    }
}